# Calculate error rates for the CV14 on-phone test (Oct 2, 2026)
Testing the Survey2026 branch build `feature/swahili-whisper-cv14` (`e0b5b6b`), built locally so the APK contains the real Swahili Whisper model `dmusingu/WHISPER-SMALL-SWAHILI-ASR-CV-14` (`ggml-small-sw-cv14-q5_0.bin`). The CI preview APK for this branch contains only a Git LFS pointer in place of the model.

Device: Samsung SM-S931U1. Survey `391b8e25`, run on Oct 2, 2026 by Frank.

Ednah's ten recordings from Sep 21 office testing (`7db59ab6`, last take per question) were played through a speaker into the phone during the survey.

In [1]:
import json

In [2]:
#define the path to the reference text
# One entry per survey recording: survey_id, question_id, expected_answer, wav_file
reference_file = "office_testing/expected_answers.json"
reference = reference_file

# Read from a relative path
with open(reference, 'r', encoding='utf-8') as g:
    ref_data = json.load(g)

# Look up the expected answer by (survey_id, question_id), since a survey can have its own reference text
ref_lookup = {(r["survey_id"], r["question_id"]): r["expected_answer"] for r in ref_data}

print (f"Reference data: loaded {len(ref_data)} entries from {reference_file}")

Reference data: loaded 40 entries from office_testing/expected_answers.json


In [3]:
# Define the GitHub URLs of the surveys we want to test today.
# Paste the normal github.com "blob" links; they are converted to raw file URLs below
survey_urls = [
    # Oct 2: CV14 branch build (e0b5b6b) on Frank's phone, replaying Ednah's 7db59ab6 recordings
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-10-02/exports/2026-10-02_16-19-27_survey_SM-S931U1_5F8C58FB249A_391b8e25-cd79-427f-8bc8-4bdb4d4c5437.json",
]

from urllib.request import urlopen

# One entry per survey: file name plus the reference/answer pairs that matched.
surveys = []
for survey_url in survey_urls:
    raw_url = survey_url.replace("https://github.com/", "https://raw.githubusercontent.com/", 1).replace("/blob/", "/", 1)
    with urlopen(raw_url) as f:
        survey_data = json.load(f)

    survey_file = raw_url.rsplit("/", 1)[-1]
    print(f"\nLoaded data from {survey_file}")

    # Free text entered for the session (e.g. who tested and when), if present.
    session_free_text = survey_data.get('meta', {}).get('session_free_text', '')
    print(f"Session free text: {session_free_text}")

    # The app's build (git commit) is recorded in exports from October onwards.
    build = survey_data.get("build", "unknown")
    print(f"Build: {build}")

    # Keep only the questions that have a reference answer for this survey, in question order.
    survey_id = survey_data['survey_id']
    ref_questions = sorted((q for (sid, q) in ref_lookup if sid == survey_id), key=lambda q: int(q[1:]))
    if not ref_questions:
        print(f"No reference entries for survey {survey_id}, skipped")
        continue
    questions = [q for q in ref_questions if q in survey_data['answers']]
    missing = [q for q in ref_questions if q not in survey_data['answers']]
    if missing:
        print(f"Not in survey, skipped: {missing}")

    references = [ref_lookup[(survey_id, q)] for q in questions]
    answers = [survey_data['answers'][q]['answer'] for q in questions]
    print(f"Found {len(answers)} transcribed answers matching the reference.")
    for q, ref, ans in zip(questions, references, answers):
        print(f"{q:>3}  expected:   {ref}")
        print(f"     transcript: {ans}")

    surveys.append({"file": survey_file, "survey_id": survey_id, "build": build,
                    "questions": questions, "references": references, "answers": answers})


Loaded data from 2026-10-02_16-19-27_survey_SM-S931U1_5F8C58FB249A_391b8e25-cd79-427f-8bc8-4bdb4d4c5437.json
Session free text: frank testing cv14 build with swahili dictation
Build: e0b5b6b
Found 10 transcribed answers matching the reference.
 Q8  expected:   Ndiyo, iliharibu majani na kupunguza mavuno.
     transcript: ndiyo iliaribu majani na kukumbuza mamono
 Q9  expected:   Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
     transcript: naweza ukubali kupoteza hadi aslimia kumi ya mabuno
Q10  expected:   Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina.
     transcript: ikiwa uharibifu unafika karibu asilimia thelathini nitabadilisha aina
Q11  expected:   Iwe na mavuno mengi na iwe inastahimili ukame na wadudu.
     transcript: iwe na mavuno mengi na iwe inastahimili ukame na wadodo
Q12  expected:   Hata tani moja kwa ekari, kama mbegu ni nzuri.
     transcript: hata tani moja kwa hekali kama mbegu ni mzuri
Q13  expected:   Wakati wa maua na kutengen

In [4]:
import jiwer
import re
from jiwer.transforms import AbstractTransform

# Swahili number words, so "30", "30%" and "thelathini" all score the same.
_UNITS = ["", "moja", "mbili", "tatu", "nne", "tano", "sita", "saba", "nane", "tisa"]
_TENS = ["", "kumi", "ishirini", "thelathini", "arobaini", "hamsini", "sitini", "sabini", "themanini", "tisini"]

def swahili_number(n):
    """Spell out 0-999,999 in Swahili, e.g. 30 -> 'thelathini', 25 -> 'ishirini na tano'."""
    if n == 0:
        return "sifuri"
    parts = []
    if n >= 1000:
        parts.append("elfu " + swahili_number(n // 1000)); n %= 1000
    if n >= 100:
        parts.append("mia " + swahili_number(n // 100)); n %= 100
    if n >= 10:
        parts.append(_TENS[n // 10]); n %= 10
    if n:
        parts.append(_UNITS[n])
    return " na ".join(parts)

class NormalizeSwahiliNumbers(AbstractTransform):
    """Digits to Swahili words: '2-3' -> 'mbili hadi tatu', '10%' -> 'asilimia kumi'."""
    def process_string(self, s):
        num = lambda m: swahili_number(int(m))
        s = re.sub(r"\b(\d+)\s*[-–]\s*(\d+)\b", lambda m: f"{num(m[1])} hadi {num(m[2])}", s)  # ranges
        s = re.sub(r"\b(?:asilimia\s+)?(\d+)\s*%", lambda m: f"asilimia {num(m[1])}", s)       # percentages
        return re.sub(r"\b\d+\b", lambda m: num(m[0]), s)                                      # plain numbers


transform_words = jiwer.Compose ([
    jiwer.ToLowerCase(),
    NormalizeSwahiliNumbers(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfWords()
])

transform_chars = jiwer.Compose ([
    jiwer.ToLowerCase(),
    NormalizeSwahiliNumbers(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfChars()
])

def error_rates(references, answers):
    out = jiwer.process_words(references, answers, reference_transform=transform_words, hypothesis_transform=transform_words)
    out2 = jiwer.process_characters(references, answers, reference_transform=transform_chars, hypothesis_transform=transform_chars)
    return out.wer, out2.cer

# Rates for each survey, then per question, then the total across all surveys.
for s in surveys:
    wer, cer = error_rates(s["references"], s["answers"])
    print(f"{s['file']}  (build {s['build']}, {len(s['answers'])} answers)")
    print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")
    print("  Per question:")
    for q, ref, ans in zip(s["questions"], s["references"], s["answers"]):
        qwer, qcer = error_rates([ref], [ans])
        print(f"    {q:>3}  WER {qwer:.3f}  CER {qcer:.3f}")

all_references = [r for s in surveys for r in s["references"]]
all_answers = [a for s in surveys for a in s["answers"]]
wer, cer = error_rates(all_references, all_answers)
print(f"\nAggregate over {len(surveys)} survey(s), {len(all_answers)} answers")
print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

2026-10-02_16-19-27_survey_SM-S931U1_5F8C58FB249A_391b8e25-cd79-427f-8bc8-4bdb4d4c5437.json  (build e0b5b6b, 10 answers)
  WER (normalized): 0.250   CER (normalized): 0.050
  Per question:
     Q8  WER 0.500  CER 0.143
     Q9  WER 0.375  CER 0.057
    Q10  WER 0.000  CER 0.000
    Q11  WER 0.100  CER 0.036
    Q12  WER 0.222  CER 0.068
    Q13  WER 0.167  CER 0.029
    Q14  WER 0.100  CER 0.052
    Q15  WER 0.273  CER 0.033
    Q16  WER 0.125  CER 0.038
    Q17  WER 0.583  CER 0.068

Aggregate over 1 survey(s), 10 answers
  WER (normalized): 0.250   CER (normalized): 0.050


## Results

CV14 on the phone: **WER 0.250, CER 0.050** across the ten answers.

| Recordings | Model | Where it ran | WER | CER |
|---|---|---|---|---|
| Ednah `7db59ab6`, original files | Build #70 (generic `ggml-small-q5_1`) | Phone, Sep 21 | 0.727 | 0.120 |
| Ednah `7db59ab6`, original files | CV14 | Cloud CPU (whisper.cpp v1.9.3) | 0.114 | 0.024 |
| **Ednah `7db59ab6`, replayed into the phone** | **CV14** | **Phone, Oct 2** | **0.250** | **0.050** |

The first two rows come from the model comparison on the `nairobi_office_testing` branch.

On the phone, CV14 cuts word errors by about two-thirds and character errors by more than half compared with build #70 on the same recordings. It is behind CV14 on the original files, which is expected: replaying through a speaker into the phone microphone loses some audio quality.

### Per question

- **Q10:** transcribed exactly.
- **Q8** (WER 0.500): "kupunguza mavuno" came out as "kukumbuza mamono". The cloud run on the original file got both words right.
- **Q17** (WER 0.583): mostly word-boundary errors ("waeneo", "kwasababu", "wakukaribu" for "wa eneo", "kwa sababu", "wako karibu"), so its CER is only 0.068.
- **Q9, Q12, Q15:** single-sound slips ("ukubali", "aslimia", "mabuno"; "hekali", "mzuri"; "na wapa", "maindi").
- **Q11, Q13, Q14, Q16:** one wrong word each ("wadodo", "punja", missing "ni", "agrovate").

Four transcripts (Q10, Q11, Q13, Q16) are identical to CV14's output on Ednah's original files, which confirms the phone build is running the CV14 model.

### Notes

- **References:** Ednah's entries in `office_testing/expected_answers.json` (the scripted answers), copied to survey `391b8e25`.
- **Normalization:** lowercase, numbers converted to Swahili words, punctuation removed, before scoring (`NormalizeSwahiliNumbers`).
- **Missing uploads:** Q11, Q14 and Q17 recordings are listed in the export but weren't uploaded to SurveyExports. Their transcripts are in the export, so all ten are scored.